shared/src/main/scala/io/estatico/newtype/macros/NewTypeMacros.scala

```diff
@@ -185,14 +185,14 @@ private[macros] class NewTypeMacros(val c: blackbox.Context)
     clsDef: ClassDef, valDef: ValDef, tparamsNoVar: List[TypeDef], tparamNames: List[TypeName]
   ): List[Tree] = {
     if (!unapply) Nil else {
-      // Note that our unapply method should Some since its isEmpty/get is constant.
+      // We are using a generic value class name-based extractor not to allocate Some
       List(
         if (tparamsNoVar.isEmpty) {
-          q"""def unapply(x: ${clsDef.name}): Some[${valDef.tpt}] =
-              Some(x.asInstanceOf[${valDef.tpt}])"""
+          q"""def unapply(x: ${clsDef.name}): io.estatico.newtype.NewType.NewTypeUnapplyOps[${valDef.tpt}] =
+                new io.estatico.newtype.NewType.NewTypeUnapplyOps[${valDef.tpt}](x.coerce)"""
         } else {
-          q"""def unapply[..$tparamsNoVar](x: ${clsDef.name}[..$tparamNames]): Some[${valDef.tpt}] =
-              Some(x.asInstanceOf[${valDef.tpt}])"""
+          q"""def unapply[..$tparamsNoVar](x: ${clsDef.name}[..$tparamNames]): io.estatico.newtype.NewType.NewTypeUnapplyOps[${valDef.tpt}] =
+                new io.estatico.newtype.NewType.NewTypeUnapplyOps[${valDef.tpt}](x.coerce[${valDef.tpt}])"""
```

Comment:
Should probably avoid the use of the `.coerce` extension method as it's unreliable from this context (requires the user to have the Coercible ops imported). Instead, should be able to use `asInstanceOf` safely.